File: php/tests/probe.php
<?php
declare(strict_types=1);
require dirname(__DIR__) . '/src/OrderedJson.php';
use OrderedJson\{Value, Member, ParseError};
use function OrderedJson\parse;

$expectNative = in_array('--native', $argv, true);
if ($expectNative !== extension_loaded('ordered_json')) {
    throw new RuntimeException('Unexpected PHP parser backend');
}

// No test cases or expectations here; the common verifier supplies documents.
function quote(string $s): string { return json_encode($s, JSON_THROW_ON_ERROR); }
function text(Value $v): string { return Value::fromUnits($v->stringUnits())->raw(); }
function tree(Value $v): string {
    return match ($v->kind()) {
        'object' => '["object",['.implode(',', array_map(
            fn(Member $m) => '['.text($m->key).','.tree($m->value).']', $v->members())).']]',
        'array' => '["array",['.implode(',', array_map(tree(...), $v->items())).']]',
        'string' => '["string",'.text($v).']',
        'number' => '["number",'.quote($v->numberLiteral()).']',
        'boolean' => '["boolean",'.($v->booleanValue() ? 'true':'false').']',
        default => '["null"]',
    };
}
function rebuild(Value $v): Value {
    return match ($v->kind()) {
        'object' => Value::object(array_map(fn(Member $m) => new Member($m->key, rebuild($m->value)), $v->members())),
        'array' => Value::array(array_map(rebuild(...), $v->items())),
        default => $v,
    };
}
while (($line = fgets(STDIN)) !== false) {
    $source = file_get_contents(rtrim($line, "\r\n"));
    if ($source === false) throw new RuntimeException('Cannot read document');
    try { $value = parse($source); }
    catch (ParseError $e) { echo '{"ok":false}', "\n"; continue; }
    echo '{"ok":true,"raw":',quote($value->raw()),',"compact":',quote($value->compact()),
        ',"tree":',tree($value),',"rebuilt":',quote(rebuild($value)->compact()),"}\n";
}

File: rust/examples/probe.rs
// Adapter only. The shared Python verifier owns examples and expectations.
use ordered_json::{parse_bytes, Kind, Member, Value};
use std::{
    fs,
    io::{self, BufRead},
};

fn quote(s: &str) -> String {
    Value::string(s).raw().to_owned()
}
fn text(v: &Value) -> String {
    Value::from_units(v.string_units().unwrap())
        .raw()
        .to_owned()
}
fn tree(v: &Value) -> String {
    match v.kind() {
        Kind::Object => format!(
            "[\"object\",[{}]]",
            v.members()
                .unwrap()
                .iter()
                .map(|m| format!("[{},{}]", text(&m.key), tree(&m.value)))
                .collect::<Vec<_>>()
                .join(",")
        ),
        Kind::Array => format!(
            "[\"array\",[{}]]",
            v.items()
                .unwrap()
                .iter()
                .map(tree)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Kind::String => format!("[\"string\",{}]", text(v)),
        Kind::Number => format!("[\"number\",{}]", quote(v.number_literal().unwrap())),
        Kind::Boolean => format!("[\"boolean\",{}]", v.boolean_value().unwrap()),
        Kind::Null => "[\"null\"]".into(),
    }
}
fn rebuild(v: &Value) -> Value {
    match v.kind() {
        Kind::Object => Value::object(
            &v.members()
                .unwrap()
                .iter()
                .map(|m| Member {
                    key: m.key.clone(),
                    value: rebuild(&m.value),
                })
                .collect::<Vec<_>>(),
        )
        .unwrap(),
        Kind::Array => {
            Value::array(&v.items().unwrap().iter().map(rebuild).collect::<Vec<_>>()).unwrap()
        }
        _ => v.clone(),
    }
}
fn main() {
    for line in io::stdin().lock().lines() {
        let bytes = fs::read(line.unwrap()).expect("cannot read document");
        match parse_bytes(&bytes) {
            Err(_) => println!("{{\"ok\":false}}"),
            Ok(v) => println!(
                "{{\"ok\":true,\"raw\":{},\"compact\":{},\"tree\":{},\"rebuilt\":{}}}",
                quote(v.raw()),
                quote(&v.compact()),
                tree(&v),
                quote(&rebuild(&v).compact())
            ),
        }
    }
}
